Chown images after build with sudo

ArangoGutierrez opened the pull request:
- Fixes #2102

Signed-off-by: ArangoGutierrez <eduardo@sylabs.io>

> test on RHEL7 - Go 1.11
```bash
[eduardo@rhel7 singularity]$ sudo singularity build test.sif examples/busybox/Singularity
WARNING: Authentication token file not found : Only pulls of public images will succeed
INFO:    Starting build...
<...>
[eduardo@rhel7 singularity]$ ls -lh test.sif
-rwxr-xr-x. 1 eduardo eduardo 640K Sep 27 10:46 test.sif
```
**Checkoff for all PRs:**

- [x] I have read the [Guidelines for Contributing](https://github.com/sylabs/singularity/blob/master/CONTRIBUTING.md), and this PR conforms to the stated requirements.
- [ ] I have added changes to the [CHANGELOG](https://github.com/sylabs/singularity/blob/development/CHANGELOG.md) and and documentation updates to the [user](https://www.github.com/sylabs/singularity-userdocs) and [administrator](https://www.github.com/sylabs/singularity-admindocs) documentation bases.
- [x] I have tested this PR locally with a `make testall`
- [x] This PR is against the project's master branch
- [x] I have added myself as a contributor to the [contributors's file](https://github.com/sylabs/singularity/blob/master/CONTRIBUTORS.md)
- [x] This PR is ready for review and/or merge


Attn: @singularity-maintainers

ArangoGutierrez committed:
Chown images after build with sudo

 - Fixes #2102

Signed-off-by: ArangoGutierrez <eduardo@sylabs.io>

bauerm97 commented on src/pkg/build/assemblers/assembler_sif.go:
```diff
@@ -77,6 +79,14 @@ func createSIF(path string, definition []byte, squashfile string) (err error) {
 
+	// chown the sif file to the calling user
+	if syscall.Getuid() == 0 {
+		uID, _ := strconv.Atoi(os.Getenv("SUDO_UID"))
```
We shouldn't be ignoring errors here, *especially* since this is touching security

bauerm97 replied:
Also, I think `uid` looks better than `uID`

ArangoGutierrez replied:
ummm lint on my IDE, recommend id always should be ID.
but I agree with you

bauerm97 commented on src/pkg/build/assemblers/assembler_sif.go:
```diff
@@ -77,6 +79,14 @@ func createSIF(path string, definition []byte, squashfile string) (err error) {
 	}
 
+	// chown the sif file to the calling user
+	if syscall.Getuid() == 0 {
```
What about the case where the calling user is root. Not running `sudo ...` but they are logged in as the root user? In this case, `$SUDO_UID` may not be set, and since we're not checking for `Atoi` errors, we can't be guaranteed of the behavior

bauerm97 reviewed (changes requested):
In general, I'm not a fan of the approach of automatically calling `chown` on an image built with `uid=0`. This seems unexpected to me, especially because we're not handling the case where the user is actually logged in as root, and not just running as sudo. 

I think this feature would be better served by implementing a flag to optionally do this. Something like `--chown <uid>:<gid>` makes much more sense to me, as it means we won't have any unexpected default behavior. @gmkurtzer I'd like to hear your thoughts here since this was your idea IIRC

ArangoGutierrez committed:
Handle case when calling user is root

Better handle the case where the calling user is root. Not a user running sudo,but they are logged in as the root user.
In this case, $SUDO_UID may not be set, and since we're not checking for Atoi error.

Signed-off-by: ArangoGutierrez <eduardo@sylabs.io>

jmstover commented on src/pkg/build/assemblers/assembler_sif.go:
```diff
@@ -77,6 +79,14 @@ func createSIF(path string, definition []byte, squashfile string) (err error) {
 	}
 
+	// chown the sif file to the calling user
+	if syscall.Getuid() == 0 && os.Getenv("USERNAME") != "root" {
```
If I run `sudo` ... `$USERNAME` is set to `root`. 

Check `SUDO_USER` instead

ArangoGutierrez committed:
Merge branch 'master' of https://github.com/singularityware/singularity into chown_builds

Signed-off-by: ArangoGutierrez <eduardo@sylabs.io>

ArangoGutierrez committed:
check  USERNAME --> SUDO_USER

Signed-off-by: ArangoGutierrez <eduardo@sylabs.io>

jmstover reviewed (approved):
LGTM

ArangoGutierrez committed:
Update Branch to upstream master

Signed-off-by: ArangoGutierrez <eduardo@sylabs.io>

ArangoGutierrez committed:
move chown bits to a local func changeOwner()

changeOwner check the command being called with sudo with the environment
variable SUDO_COMMAND. Pattern match that for the singularity bin

Signed-off-by: ArangoGutierrez <eduardo@sylabs.io>

ArangoGutierrez committed:
Merge branch 'master' of https://github.com/singularityware/singularity into chown_builds

bauerm97 commented on src/pkg/build/assemblers/assembler_sif.go:
```diff
@@ -129,3 +136,41 @@ func (a *SIFAssembler) Assemble(b *types.Bundle, path string) (err error) {
+			}
+			gid, err = strconv.Atoi(_gid)
+			if err != nil {
+				sylog.Errorf("Error while calling strconv : %v", err)
```
I would say `Warningf` since these are not "errors" which will cause singularity to exit

bauerm97 replied:
Same as the other `Errorf` spots as well

bauerm97 commented on src/pkg/build/assemblers/assembler_sif.go:
```diff
@@ -129,3 +136,41 @@ func (a *SIFAssembler) Assemble(b *types.Bundle, path string) (err error) {
+
+	r := regexp.MustCompile("(singularity)")
+	sudoCmd := os.Getenv("SUDO_COMMAND")
+	if r.MatchString(sudoCmd) {
```
```
if !r.MatchString(sudoCmd) {
    return 0, 0, false
}
```

bauerm97 commented on src/pkg/build/assemblers/assembler_sif.go:
```diff
@@ -129,3 +136,41 @@ func (a *SIFAssembler) Assemble(b *types.Bundle, path string) (err error) {
+// changeOwner check the command being called with sudo with the environment
+// variable SUDO_COMMAND. Pattern match that for the singularity bin
+func changeOwner() (int, int, bool) {
+	var uid, gid int
```
Can probably omit this here and just use `:=` syntax to declare these vars at appropriate times

bauerm97 commented on src/pkg/build/assemblers/assembler_sif.go:
```diff
@@ -129,3 +136,41 @@ func (a *SIFAssembler) Assemble(b *types.Bundle, path string) (err error) {
+	r := regexp.MustCompile("(singularity)")
+	sudoCmd := os.Getenv("SUDO_COMMAND")
+	if r.MatchString(sudoCmd) {
+		if syscall.Getuid() == 0 && os.Getenv("SUDO_USER") != "root" {
```
```
if !(syscall.Getuid() == 0 && os.Getenv("SUDO_USER") != "root") {
    return 0, 0, false
}
```

bauerm97 reviewed (changes requested):
In general pretty close to being ready. I think we should typically prefer the design pattern of avoiding lots of nested conditionals as it's harder to read. Also we should reserve `Errorf` for when we're actually returning an `error` type. If we're catching an error but are not worried about returning the error, we should be using `Warningf` instead.

ArangoGutierrez committed:
break nested conditionals and use Warningf

avoiding lots of nested conditionals as it's harder to read.
Also we should reserve Errorf for when we're actually returning an error type.
If we're catching an error but are not worried about returning the error, we should be using Warningf instead.

Signed-off-by: ArangoGutierrez <eduardo@sylabs.io>

bauerm97 commented on src/pkg/build/assemblers/assembler_sif.go:
```diff
@@ -129,3 +136,41 @@ func (a *SIFAssembler) Assemble(b *types.Bundle, path string) (err error) {
+
+		return 0, 0, false
+	}
+	chown := true
```
This line doesn't serve a purpose. Just `return uid, gid, true` is okay.

ArangoGutierrez replied:
agree

bauerm97 commented on src/pkg/build/assemblers/assembler_sif.go:
```diff
@@ -129,3 +136,40 @@ func (a *SIFAssembler) Assemble(b *types.Bundle, path string) (err error) {
+		return 0, 0, false
+	}
+
+	if syscall.Getuid() == 0 && os.Getenv("SUDO_USER") == "root" {
```
This logic here is incorrect.

First of all, we want to continue only if the UID == 0. Second, if `SUDO_USER` is set to *anything* we should continue.

I think the logic should look like this:
`if os.Getenv("SUDO_USER") == "" || syscall.Getuid() != 0 {...}`

This means we *only* proceed with the function *if and only if* `SUDO_USER` is *set* and `uid==0`.

ArangoGutierrez replied:
066de5d

ArangoGutierrez committed:
Fix if statement logic on line 149

if os.Getenv("SUDO_USER") == "" || syscall.Getuid() != 0 {...}
This means we only proceed with the function if and only if SUDO_USER is set and uid==0.

Signed-off-by: ArangoGutierrez <eduardo@sylabs.io>

bauerm97 commented on src/pkg/build/assemblers/assembler_sif.go:
```diff
@@ -81,6 +83,11 @@ func createSIF(path string, definition []byte, squashfile string) (err error) {
 
+	// chown the sif file to the calling user
+	if uid, gid, ok := changeOwner(); ok {
+		os.Chown(path, uid, gid)
```
We should always be checking for errors from functions which return errors

ArangoGutierrez replied:
agree.
done in 
2d22e57

ArangoGutierrez committed:
check for err on os.Chown()

Signed-off-by: ArangoGutierrez <eduardo@sylabs.io>

ArangoGutierrez committed:
fix merge conflicts

Signed-off-by: ArangoGutierrez <eduardo@sylabs.io>

ArangoGutierrez committed:
Merge branch 'master' into chown_builds

cclerget reviewed (approved):
Looks good !

ArangoGutierrez committed:
Merge branch 'master' into chown_builds
